File: apps/web/src/lib/tenancy.ts
/**
 * apps/web/src/lib/tenancy.ts
 * ============================
 * Identidade de tenant verificada — a contraparte em Next.js de
 * agents/cmo_agent/tenancy.py e agents/pipeline/shared/tenancy.py.
 *
 * Os três módulos leem o MESMO documento Firestore `tenants/{id}` e o MESMO
 * segredo `TENANT_KEY_PEPPER`, então uma chave emitida para um tenant vale
 * nos três serviços sem sincronização adicional.
 *
 * Regra (idêntica nos três lugares):
 *   - tenantId ausente ou "default" → tenant único implícito (o operador de
 *     hoje), SEM chave exigida. É o comportamento atual, preservado.
 *   - qualquer outro tenantId → precisa de X-Tenant-Key válida, verificada
 *     por HMAC contra `tenants/{id}.key_hash`.
 *
 * Antes deste módulo, toda rota que lia `request.headers.get('x-tenant-id')`
 * confiava nele sem checagem — qualquer chamador autenticado pelo cookie
 * único do CSM podia se declarar dono de qualquer tenant. Sem risco prático
 * enquanto só existe um tenant, mas deixa de ser inofensivo no dia em que um
 * segundo existir.
 */

import { createHmac, timingSafeEqual } from 'node:crypto';
import { getFirestoreDb } from '@/lib/firebase';

export const DEFAULT_TENANT_ID = 'default';
const COLLECTION_TENANTS = 'tenants';

export interface TenantContext {
  tenantId: string;      // 'default' para o operador único
  name: string;
  status: string;
  isDefault: boolean;
}

export type ResolveTenantResult =
  | { ok: true; tenant: TenantContext }
  | { ok: false; status: number; error: string };

function pepper(): string {
  return (process.env.TENANT_KEY_PEPPER || '').trim();
}

/**
 * HMAC-SHA256 salgado pelo tenantId — idêntico ao hash_tenant_key() dos dois
 * módulos Python. Vazar o hash de um tenant não ajuda a forjar a chave de
 * outro, porque o salt é o tenantId, não um valor fixo global.
 */
export function hashTenantKey(tenantId: string, rawKey: string): string {
  const p = pepper();
  if (!p) {
    throw new Error(
      'TENANT_KEY_PEPPER não configurado — necessário para verificar chaves de tenant não-default.',
    );
  }
  return createHmac('sha256', p).update(`${tenantId}:${rawKey}`).digest('hex');
}

function timingSafeEqualHex(a: string, b: string): boolean {
  const bufA = Buffer.from(a, 'hex');
  const bufB = Buffer.from(b, 'hex');
  return bufA.length === bufB.length && timingSafeEqual(bufA, bufB);
}

/**
 * Resolve e verifica a identidade do tenant a partir dos headers da
 * requisição. Não lança — devolve um Result para o chamador decidir o
 * HTTP status (as rotas de API já padronizam NextResponse.json em cada
 * handler, então uma exceção aqui obrigaria try/catch em toda rota).
 */
export async function resolveTenant(request: Request): Promise<ResolveTenantResult> {
  const rawTenantId = request.headers.get('x-tenant-id');
  const tenantKey   = request.headers.get('x-tenant-key');

  const tid = (rawTenantId || '').trim() || DEFAULT_TENANT_ID;

  if (tid === DEFAULT_TENANT_ID) {
    // Comportamento de hoje, preservado: não consulta o Firestore, não exige
    // chave — é o único operador, sem outro tenant para se confundir com.
    return {
      ok: true,
      tenant: { tenantId: DEFAULT_TENANT_ID, name: 'éozoré', status: 'active', isDefault: true },
    };
  }

  const db = getFirestoreDb();
  if (!db) {
    return { ok: false, status: 503, error: 'Firestore indisponível para verificar tenant' };
  }

  const snap = await db.collection(COLLECTION_TENANTS).doc(tid).get();
  if (!snap.exists) {
    return { ok: false, status: 403, error: `Tenant '${tid}' não existe` };
  }

  const data = snap.data() ?? {};
  if (data.status !== 'active') {
    return { ok: false, status: 403, error: `Tenant '${tid}' não está ativo (status=${data.status})` };
  }

  const storedHash = String(data.key_hash || '');
  if (!tenantKey || !storedHash) {
    return { ok: false, status: 403, error: `Tenant '${tid}' exige X-Tenant-Key` };
  }

  let presentedHash: string;
  try {
    presentedHash = hashTenantKey(tid, tenantKey);
  } catch (err) {
    return { ok: false, status: 500, error: err instanceof Error ? err.message : String(err) };
  }

  if (!timingSafeEqualHex(presentedHash, storedHash)) {
    return { ok: false, status: 403, error: `Chave inválida para tenant '${tid}'` };
  }

  return {
    ok: true,
    tenant: {
      tenantId: tid,
      name: String(data.name || tid),
      status: data.status,
      isDefault: false,
    },
  };
}

/**
 * Açúcar para rotas que só precisam do tenantId no formato que dbPaths já
 * espera (`string | null`, null = tenant default / coleções raiz).
 *
 * Devolve `Response` pronta para `return` quando a verificação falha —
 * mantém as rotas de API com uma linha a mais, não um bloco de tratamento.
 */
export async function requireTenantId(
  request: Request,
): Promise<{ tenantId: string | null } | { response: Response }> {
  const result = await resolveTenant(request);
  if (!result.ok) {
    return {
      response: new Response(JSON.stringify({ error: result.error }), {
        status: result.status,
        headers: { 'Content-Type': 'application/json' },
      }),
    };
  }
  return { tenantId: result.tenant.isDefault ? null : result.tenant.tenantId };
}

File: agents/pipeline/shared/tenancy.py
"""
agents/pipeline/shared/tenancy.py
==================================
Identidade de tenant verificada — o que faltava para o `tenant_id` que já
circula pela pipeline (Firestore em `tenants/{id}/...`, `TENANT_ID` env var,
header `X-Tenant-ID`) significar alguma coisa.

Hoje, `tenant_id` é aceito de qualquer chamador sem checagem nenhuma: qualquer
requisição autenticada pelo segredo único do CSM pode se declarar dono de
qualquer tenant só passando o header. Não é um problema explorável enquanto só
existe um tenant ("default"), mas deixa de ser inofensivo no dia em que um
segundo tenant existir — e é exatamente esse dia que este módulo prepara, sem
mexer em nada do que já funciona hoje.

Regra:
  - tenant_id ausente ou "default" → tenant implícito único, SEM chave exigida.
    É o comportamento atual, preservado byte a byte.
  - qualquer outro tenant_id → precisa de uma chave válida (X-Tenant-Key),
    verificada por HMAC contra o hash salvo em `tenants/{id}.key_hash`.
    Sem isso, ou com tenant status != "active", a requisição é rejeitada.

Este módulo cobre IDENTIDADE. Orçamento mensal (quanto o tenant já gastou em
HeyGen/ElevenLabs este mês) é responsabilidade de cost_tracker.py — as duas
coisas são independentes: um tenant pode estar autenticado e ainda assim
bloqueado por orçamento.
"""

from __future__ import annotations

import hashlib
import hmac
import logging
import os
import secrets
from dataclasses import dataclass
from typing import Optional

from google.cloud.firestore_v1 import AsyncClient

logger = logging.getLogger(__name__)

COLLECTION_TENANTS = "tenants"
DEFAULT_TENANT_ID = "default"

# Pepper do servidor para o HMAC da chave de tenant — nunca a chave em si.
# Mesma família do CSM_AUTH_SECRET (apps/web/src/lib/csmAuth.ts): um segredo
# guardado só no backend, comparado por HMAC, nunca reversível.
_PEPPER_ENV = "TENANT_KEY_PEPPER"


class TenantAuthError(Exception):
    """Tenant não encontrado, chave inválida, ou tenant suspenso."""


@dataclass(frozen=True)
class TenantContext:
    tenant_id: str
    name: str
    status: str

    @property
    def is_default(self) -> bool:
        return self.tenant_id == DEFAULT_TENANT_ID


def _pepper() -> str:
    value = os.environ.get(_PEPPER_ENV, "").strip()
    if not value:
        # Falha alto em vez de silenciosamente aceitar qualquer chave — um
        # pepper vazio tornaria hash_tenant_key determinístico e adivinhável.
        raise RuntimeError(
            f"{_PEPPER_ENV} não configurado. Necessário para verificar chaves "
            "de tenant que não sejam o tenant default."
        )
    return value


def hash_tenant_key(tenant_id: str, raw_key: str) -> str:
    """
    HMAC-SHA256 da chave, salgado pelo tenant_id — o mesmo texto de chave gera
    hashes diferentes por tenant, então vazar o hash de um tenant não ajuda a
    forjar a chave de outro.
    """
    msg = f"{tenant_id}:{raw_key}".encode("utf-8")
    return hmac.new(_pepper().encode("utf-8"), msg, hashlib.sha256).hexdigest()


def generate_tenant_key() -> str:
    """Chave aleatória para um novo tenant — mostrada UMA vez no bootstrap."""
    return secrets.token_urlsafe(32)


async def resolve_tenant(
    db: AsyncClient,
    tenant_id: Optional[str],
    tenant_key: Optional[str],
) -> TenantContext:
    """
    Resolve e verifica a identidade do tenant para uma requisição.

    Raises:
        TenantAuthError: tenant_id != default sem chave válida, tenant
            desconhecido, ou tenant com status != "active".
    """
    tid = (tenant_id or "").strip() or DEFAULT_TENANT_ID

    if tid == DEFAULT_TENANT_ID:
        # Comportamento atual, sem exigir nada novo: é o único operador de
        # hoje, e forçar uma chave aqui quebraria toda chamada existente.
        return TenantContext(tenant_id=DEFAULT_TENANT_ID, name="éozoré", status="active")

    doc = await db.collection(COLLECTION_TENANTS).document(tid).get()
    if not doc.exists:
        raise TenantAuthError(f"Tenant '{tid}' não existe")

    data = doc.to_dict() or {}
    if data.get("status") != "active":
        raise TenantAuthError(f"Tenant '{tid}' não está ativo (status={data.get('status')})")

    stored_hash = data.get("key_hash", "")
    if not tenant_key or not stored_hash:
        raise TenantAuthError(f"Tenant '{tid}' exige X-Tenant-Key")

    presented_hash = hash_tenant_key(tid, tenant_key)
    # Comparação em tempo constante — hmac.compare_digest, não `==`, para não
    # vazar por timing quantos bytes do hash bateram.
    if not hmac.compare_digest(presented_hash, stored_hash):
        raise TenantAuthError(f"Chave inválida para tenant '{tid}'")

    return TenantContext(
        tenant_id=tid,
        name=data.get("name", tid),
        status=data["status"],
    )
